### PR TITLE
fix: archived reports are unable to mark as read

### PR DESCRIPTION
**Pull request by hurali97**
### Details

This fixes the issue where the user is unable to mark an archived chat as read by switching to it. Issue is that we check whether the `newMessageTimeReference` is lesser than the `reportAction.created` but in the case of archived reports, we have `newMessageTimeReference` equal to `reportAction.created`. So the fix is to update the condition to account for equal to and also future proof of any regression by using this new condition if the report is archived.

### Fixed Issues

$ https://github.com/Expensify/App/issues/45228
PROPOSAL: https://expensify.slack.com/archives/C05LX9D6E07/p1720520521629939?thread_ts=1720290639.217409&cid=C05LX9D6E07

### Tests

Steps:

- Go to the unread archived chat and click on it
- Verify that it's marked as read, which means bold indicator will be removed

<hr/>

- [x] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/47336142/65138270-cc8d-4b41-a95e-f64485bb7b5f

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/47336142/fec69a84-14d8-4007-a51b-21dbab7449b1

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/47336142/fcd78cdf-e997-47d5-bf9b-c0bd5f06ab7a

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/47336142/eab9bf6a-5176-4258-864c-40e4fe10b1f8

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47336142/61322e00-aa7a-468e-9832-7544f3bacd41

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/47336142/2d19b63a-aa14-48c4-a39b-3de0025d0a8c

</details>

### CODE REVIEW

**Review by mkhutornyi — CHANGES_REQUESTED** [review]
I think your solution is workaround.
This used to work before. There should be recent regression. Let's fix the root cause.

Note: Mark as unread is also not working (maybe same root cause)

https://github.com/Expensify/App/assets/97676131/372a179d-2878-4bda-a308-f28acde53514

**Comment by hurali97** [comment]
@mkhutornyi This maybe a regression. In my investigation, this only happens because of the `newMessageTimeReference` is not lesser than the `reportAction.created`. In archived chats case, `newMessageTimeReference` contains `report.lastReadTime`. Now both `reportAction.created` and `report.lastReadTime` are being populated from the backend. This message [here](https://expensify.slack.com/archives/C05LX9D6E07/p1720640487498039?thread_ts=1720290639.217409&cid=C05LX9D6E07) explains there is some recent backend changes which affect the read/unread status of archived reports.

Based on this [message](https://expensify.slack.com/archives/C05LX9D6E07/p1720657644598809?thread_ts=1720290639.217409&cid=C05LX9D6E07) we went on to create the PR because the backend changes might take some time.

<hr/>

For unable to mark as unread, I was only able to reproduce it on some archived read chats. This only happens for chats which have some sort of transactions. When we click on **Mark as unread** the below code is executed:

```ts
 const originalReportID = ReportUtils.getOriginalReportID(reportID, reportAction) ?? '-1';
 Report.markCommentAsUnread(originalReportID, reportAction?.created);
```

So this helper function `getOriginalReportID` does three things:

1. If `isThreadFirstChat` is true, it will return the `parentReportID`
2. Else, if `transactionThreadReportID` is present it will return it
3. Otherwise, the passed `reportID` will be returned

In the case of where we are unable to mark as unread, we get the `transactionThreadReportID` and use that to mark the chat as unread. Now, there are two things:

- We have `reportID` associated with the Option item in the LHN, say `123`
- We get `transactionThreadReportID` as `originalID` from `getOriginalReportID` function, say `456`
- We use the above ID, '456' to mark the chat as unread, the API succeeds but we still don't see any updates in the LHN

We don't see any update in the LHN because the `reportID` associated is different than the one we used to mark as unread. Let's see the video below, where `originalID` is not the same as `reportID`:


https://github.com/user-attachments/assets/ca808613-6059-45df-b38d-bd904ca9a323


For other archived chats we don't see this issue, in below video we get the `originalID` same as `reportID` and everything works as expected:



https://github.com/user-attachments/assets/25605eb7-b228-46a3-a70b-e35c2662a437

<hr/>

Now that we know what's the issue, we want to figure out why is that happening. From git lens, I stumbled upon this [PR](https://github.com/Expensify/App/pull/34640) which introduced the usage of `originalID` while marking a report as unread. Prior to that PR, we were using the `reportID` to mark the report as unread. So in the code, I made this change and I was still unable to mark the chat as unread. This happens because we are now passing the wrong `lastReadTime` to the API. The reason is that we have `reportAction.created` as `undefined` due to which we fallback to get the lastActionTime from other users. See the video below:


https://github.com/user-attachments/assets/835766ef-fac5-42b5-9d24-8cb735cf3f46



We can fix this by explicitly checking, if the `reportAction.created` is `undefined`, we can get the `lastVisibleActionCreated` from the `report` using the `reportID`. This fixes the issue and we are able to mark the chat as unread. 🎉 


https://github.com/user-attachments/assets/7fc39304-53c6-4692-a7f7-44a04e80f18e


<hr/>

We can also explore another solution, which is to check whether the report associated with the `originalID` is unread. If yes, we will mark the current report unread as well.

```ts
   result.isUnread = isOriginalReportUnread || ReportUtils.isUnread(report) && !!report.lastActorAccountID;
```

<hr/>

In summary, I don't think these two issues have the same root cause and must be handled separately.

**Comment by hurali97** [comment]
> "Mark as read" in LHN context menu works perfect without this PR. Why can't we apply same approach when we open report?

I am not sure you're following through the findings shared in the slack thread and here. Again, the reason is that `readNewestAction` is called when we use `Mark as read` from LHN context menu. This same function is called when some conditions are true BUT those conditions are never true in case of archived reports. In this PR we are fixing those conditions to be able to call `readNewestAction`.

**Comment by hurali97** [comment]
> @hurali97 you may think I am asking stupid question but your solution doesn't work for me at all. Please see this video:
> 
>  Screen.Recording.2024-07-18.at.12.27.53.AM.mov

@mkhutornyi Thanks for showing the video, it's easier and clearer to see the issue now. What we fixed in this PR was the case where the archived report is already unread and in that case, this PR works fine. 

Now that you're trying to manually mark it as unread and then trying to navigate to that report to mark it as read, it's a different case. I will see why this is not working, thanks 🙂

**Comment by hurali97** [comment]
@mkhutornyi I tried this and it happened only 1 archived chat. The reason is that we have `newMessageTimeReference` ( report.lastReadTime ) and `reportAction.created` just fractionally different in the milliseconds. Hence, our updated condition doesn't work which checks for the string equality. We are getting these timestamps from the backend and `report.lastReadTime` and `reportAction.created` should be lesser than or equal.

Screenshot 2024-07-18 at 12 51 43 PM


Now, I think we can adjust the condition to check if it's an archived room, we will mark it as read, so the below change is what I suggest and it works as well:

```diff

diff --git a/src/pages/home/report/ReportActionsList.tsx b/src/pages/home/report/ReportActionsList.tsx
index 74629eb0df..6ce1ab4d0f 100644
--- a/src/pages/home/report/ReportActionsList.tsx
+++ b/src/pages/home/report/ReportActionsList.tsx
@@ -547,9 +547,7 @@ function ReportActionsList({
              * So the condition should be `<=` to mark the report as read. We will only use this condition when
              * the report is archived to not introduce any regression for other reports.
              */
-            const isTimeLesserOrEqual = ReportUtils.isArchivedRoom(report)
-                ? newMessageTimeReference && newMessageTimeReference <= reportAction.created
-                : newMessageTimeReference && newMessageTimeReference < reportAction.created;
+            const isTimeLesserOrEqual = ReportUtils.isArchivedRoom(report) || (newMessageTimeReference && newMessageTimeReference < reportAction.created);
             return (
                 isTimeLesserOrEqual &&
                 (ReportActionsUtils.isReportPreviewAction(reportAction) ? reportAction.childLastActorAccountID : reportAction.actorAccountID) !== Report.getCurrentUserAccountID()

```

<hr/>





https://github.com/user-attachments/assets/c0f65983-5c89-41c6-a7ac-81a898bf05cc

**Comment by hurali97** [comment]
> Thanks. That solution looks promising. Still need to test and verify on my side. Please appy that solution and resolve conflict.
> 
> P.S. It would be good to provide clear repro step from the beginning because
> 
> > "Mark as unread" is also not working (with different root cause), hard to re-reproduce original bug once read.

Great. I mentioned the following steps in the PR description, which you can also take in order to validate whether an unread archived chat is being marked as read:

> Steps:
> 
> Go to the unread archived chat and click on it
> Verify that it's marked as read, which means bold indicator will be removed

<hr/>

I have applied the changes. Please let me know if you're able to mark an unread archived chat as read.

**Comment by hurali97** [comment]
@mkhutornyi Any updates on the review? 🙂

**Comment by hurali97** [comment]
@mkhutornyi Oh my bad. I have adjusted the condition now, we only check if the report is archived and if yes, then we mark it as read, no need for other conditions.

Please check this again and let me know, fingers crossed 🙂

File: src/pages/home/report/ReportActionsList.tsx [hunk 1]
```diff
@@ -532,24 +541,27 @@ function ReportActionsList({
...
-                    (ReportActionsUtils.isReportPreviewAction(reportAction) ? reportAction.childLastActorAccountID : reportAction.actorAccountID) !== Report.getCurrentUserAccountID(),
-            )
-        ) {
+        const areSomeReportActionsUnread = sortedVisibleReportActions.some((reportAction) => {
+            /**
+             * The archived reports should not be marked as unread. So we are checking if the report is archived or not.
+             * If the report is archived, we will mark the report as read.
+             */
+            const isArchivedReport = ReportUtils.isArchivedRoom(report);
+            const isUnread = isArchivedReport || (newMessageTimeReference && newMessageTimeReference < reportAction.created);
```
[the remark this answers is not in the document]
Comment (hurali97) [reply src/pages/home/report/ReportActionsList.tsx:550]: @mkhutornyi No this won't cause any regression. The reason is that this code will only be executed when we click on the LHN item. So if we fix the "Mark as unread" bug in future, it would mean that we will be able to right click on the LHN item and mark that as unread, which is irrespective of the changes here.

Also, since "Mark as unread" currently works for me, so I tested this PR and it works fine 👍 


https://github.com/user-attachments/assets/a406223a-247c-4db3-84c6-12f9606fba69

**Review by mkhutornyi — APPROVED** [review]

**Review by grgia — APPROVED** [review]
LGTM, thank you!
